go/consensus/tendermint/apps/governance/governance.go

```diff
@@ -0,0 +1,406 @@
+		if err != nil {
+			return fmt.Errorf("failed to set pending upgrade: %w", err)
+		}
+	case proposal.Content.CancelUpgrade != nil:
+		cancelingProposal, err := state.Proposal(ctx, proposal.Content.CancelUpgrade.ProposalID)
+		if err != nil {
+			return fmt.Errorf("failed to query proposal: %w", err)
+		}
+		if cancelingProposal.Content.Upgrade == nil {
+			return fmt.Errorf("%w: canceling proposal needs to be an upgrade proposal", governance.ErrNoSuchUpgrade)
+		}
+		_, err = state.PendingUpgradeProposal(ctx, cancelingProposal.ID)
+		if err != nil {
+			return fmt.Errorf("failed to get pending upgrade: %w", err)
+		}
+		err = state.RemovePendingUpgrade(ctx, cancelingProposal.Content.Upgrade.Epoch, cancelingProposal.ID)
+		if err != nil {
+			return fmt.Errorf("failed to remove pending upgrade: %w", err)
+		}
+	default:
+		return governance.ErrInvalidArgument
+	}
+
+	proposal.State = governance.StatePassed
+	// If successful, emit Proposal executed event.
+	evt := &governance.ProposalExecutedEvent{
+		ID: proposal.ID,
+	}
+	ctx.EmitEvent(api.NewEventBuilder(app.Name()).Attribute(KeyProposalExecuted, cbor.Marshal(evt)))
+
+	return nil
+}
+
+func (app *governanceApplication) validatorsEscrow(
+	ctx *api.Context,
+	stakingState *stakingState.MutableState,
+	registryState *registryState.MutableState,
+	schedulerState *schedulerState.MutableState,
+) (*quantity.Quantity, map[stakingAPI.Address]*quantity.Quantity, error) {
+	currentValidators, err := schedulerState.CurrentValidators(ctx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to query current validators: %w", err)
+	}
+
+	totalVotingStake := quantity.NewQuantity()
+	validatorEntitiesEscrow := make(map[stakingAPI.Address]*quantity.Quantity)
+
+	for valID := range currentValidators {
+		var node *node.Node
+		node, err = registryState.NodeBySubKey(ctx, valID)
+		// XXX: in unit tests where DebugStaticValidators is used, the node
+		// might actually not exist in registry - causing a consensus panic here.
+		// Make sure all validators are registered in governance unit tests.
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to query validator node: %w", err)
+		}
+		entityAddr := stakingAPI.NewAddress(node.EntityID)
+
+		var escrow *quantity.Quantity
+		escrow, err = stakingState.EscrowBalance(ctx, entityAddr)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to query validator escrow: %w", err)
+		}
+
+		// If there are multiple nodes in the validator set belonging to the same entity,
+		// only count the entity escrow once.
+		if validatorEntitiesEscrow[entityAddr] != nil {
+			continue
+		}
+		validatorEntitiesEscrow[entityAddr] = escrow
+		if err := totalVotingStake.Add(escrow); err != nil {
+			return nil, nil, fmt.Errorf("failed to add to totalVotingStake: %w", err)
+		}
+	}
+	return totalVotingStake, validatorEntitiesEscrow, nil
+}
+
+func (app *governanceApplication) closeProposal(
```

Comment:
Add a comment that this may modify the proposal in case an error happens.

Reply:
added a comment that the method modifies the passed proposal